### PR TITLE
itx: one WebSocket per tab — Session gate + slug-addressed itx + invisible reconnect

### PR DESCRIPTION
**Pull request by jonastemplestein**
## Three itx WebSockets → one

Loading a project in `os.iterate.com` opened **three** WebSocket connections to `/api`, all same-origin, same auth (`from-server-cookie`), same capnweb protocol. They were three only because the browser keyed sockets in a module `Map` by `[projectId, connectionKey]`:

1. a **global** socket for the sidebar catalog,
2. a **project** socket for the page,
3. an isolated **stream-mirror** socket, whose aggressive catch-up/reconnect loop had to be kept off the page socket so a mirror reset couldn't reject a concurrent `Repo.readFile` (the `/repos/iterate` "blanked the task board" incident).

The `Map` + `connectionKey` + identity-bound eviction existed **only** to keep those independent sockets from killing each other. This collapses them to **one socket per tab** and, in doing so, deletes that whole machinery — plus the `<ItxProvider>` gate and the 12 per-page suspense boundaries that were scaffolding around it.

## The model

`authenticate()` is the top noun (as the README's "four nouns" already say):

```
ONE WebSocket ─▶ authenticate({from-server-cookie}) ─▶ Session ─▶ session.projects.get(slug) ─▶ itx
```

- **`useIterateSession()` / `connectIterateSession()`** — the catalog (`projects.list/create/get`, admin `streams`). "Iterate session" to not collide with the *auth* session from `useAuthClient()`.
- **`useItx(slug?)` / `connectItx(slug)`** — a project itx, `session.projects.get(slug)`. `projects.get` now accepts a **URL slug** (or `prj_` id), so the browser passes `params.projectSlug` straight through — no client-side slug→id hop.
- **`<ProjectScope slug>`** carries the ambient slug **and** pre-warms the one socket. No provider is needed above it — the socket is module-global and every hook dials it lazily, so the sidebar / ⌘K / admin use itx with no `<ProjectScope>` at all.
- The **stream mirror rides the shared session**; `connectionKey` is gone.

### Reconnect is invisible

The old model re-suspended every consumer on reconnect (the connection hook itself was a suspense point). Now React reads an immutable `Snapshot` whose `session` is kept across a transport gap, so **`useIterateSession()` suspends exactly once (first load) and never again**. A dropped socket always re-dials — with a session in hand invisibly, without one by re-pointing `use()` to a fresh backoff-paced promise (never a wedge). Reads are stale-while-revalidate: a resolved `useItxQuery` keeps its cached data across a blip, and only an in-flight read retries (on a finite, transport-error-only policy). No spinners on reconnect.

### Transport health is socket-owned and generation-guarded

A half-open socket (laptop sleep, network switch — no `close` event) is caught by one verifier per generation (periodic + visibility/online probes, two-strike). The mirror and the subscription watchdog **report** suspicion (`reportTransportSuspicion()`); they never close the shared socket. Generation **object identity** is the compare-and-swap that stops a late verdict from closing a healthy successor. `reconnectIterateSession()` remains the separate, deliberate **semantic** reset (new claims after create/unlock).

### One session identity (second thermo round)

`authenticate()` returns capnweb's `RpcPromise` — a *thenable* — and a native promise **assimilates** a thenable it's resolved with. Publishing the raw pipelined root therefore forked the session identity (the snapshot held the RpcPromise; imperative awaiters got the pulled resolution — two stub-cache keys, doubled project stubs, half-missed disposal, and a hung `authenticate` nothing timed out). The dial now **awaits `authenticate` once and publishes the settled stub** — one identity everywhere; the dial timeout spans the whole handshake; a real auth rejection is terminal — it surfaces from the connecting promise and **parks** the failed generation (it keeps owning the slot so a render can never trigger a fresh dial; Bugbot caught that the first version of this fix cleared the slot instead, which the failure's own re-render would have turned into an unbounded socket storm — revival is explicit via `reconnectIterateSession()`). And the parked snapshot carries **no session**: keeping the last session is the invisible-reconnect move for *transport* gaps, but a terminal auth rejection is an *authority* loss — the prior session's socket is already closed and no redial revives it, so serving it would wedge hooks on zombie stubs while the real error never surfaced (Bugbot's follow-up catch). The dead session's caches are disposed on the spot. Relatedly, first-load suspense rides one **stable first-connect promise**: React replays a suspended, never-committed component against the thenable it first used, so a per-dial rejection would have hit the error boundary even though a paced re-dial was already underway.

## The hooks & components — the whole surface

The entire browser API is **15 exports from one file** ([`itx-react.tsx`](https://github.com/iterate/iterate/blob/three-websockets/apps/os/src/itx/itx-react.tsx)). The everyday four are `useIterateSession` / `useItx` / `useItxQuery` / `useLiveState`; the rest are imperative siblings, the one mount, the escape hatch, and types. `Itx` = `RpcStub<Project>`; a `slug` argument also accepts a `prj_…` id.

| Symbol | Kind | What it gives you |
| --- | --- | --- |
| `useIterateSession()` | hook | The **Session** catalog `authenticate()` returned (`projects.list/create/get`, admin `streams`). Suspends once, on first connect. |
| `useItx(slug?)` | hook | A **project itx** (`session.projects.get(slug)`). Defaults to the ambient `<ProjectScope>`. The everyday handle: `itx.streams`, `itx.repos`, `itx.chat`, … |
| `connectIterateSession()` | fn | Imperative `useIterateSession` — a `Promise`, for handlers / `mutationFn`s / closures. |
| `connectItx(slug)` | fn | Imperative `useItx` — a `Promise`. |
| `useItxQuery({ key, query })` | hook | Read once through a **project** itx; **suspends**. Content-addressed / historical / one-shot reads. |
| `useIterateSessionQuery({ key, query })` | hook | Read once through the **session**; **non-suspending** (serves the always-mounted shell — sidebar, ⌘K, admin). |
| `useLiveState(node, selector)` | hook | Subscribe to a `.liveState` node; server pushes a snapshot then diffs, the stable-slice `selector` picks what you render. Never suspends. **The LiveView primitive.** |
| `useItxSubscription(subscribe, deps)` | hook | Low-level escape hatch for raw ordered event streams (the activity tail). Most UI wants `useLiveState`. |
| `<ProjectScope slug>` | component | Sets the ambient project + pre-warms the socket. The one mount. |
| `reconnectIterateSession()` | fn | The deliberate **semantic reset** — drop + re-dial to pick up new claims. Distinct from the automatic invisible transport reconnect. |
| `reportTransportSuspicion()` | fn | Escape hatch for the non-React transport consumer (the stream mirror): "this socket looks half-open." Routes to the socket-owned verifier; never closes the socket. |
| `isItxTransportError(e)` | fn | Is an error a transport-close (retryable), vs auth/validation/app (not)? |
| `Itx` | type | `RpcStub<Project>` — a project handle, for typing helpers. |
| `ItxLiveSubscriptionHandle` | type | What any `subscribe()` returns — `ping()` + `unsubscribe()`. |
| `ItxSubscriptionStatus` | type | `"connecting" \| "live" \| "error"`. |

**On the names** (this got a hard look — `useRead` was rejected as meaningless). The hooks are a deliberate cartesian grid over **{session, project} × {handle, read, imperative}**, so the name tells you both axes:

| | handle | finite read | imperative |
| --- | --- | --- | --- |
| **session** | `useIterateSession` | `useIterateSessionQuery` | `connectIterateSession` |
| **project** | `useItx` | `useItxQuery` | `connectItx` |

We deliberately **did not** collapse the two read hooks into one `useRead({ from })`: TanStack Query splits suspending (`useSuspenseQuery`) from non-suspending (`useQuery`), so a merged hook would need an option that flips its return type — more confusing, not less. The one awkward name is `useIterateSessionQuery` (long) — the honest cost of disambiguating from the auth session; happy to shorten it if you'd rather.

### In practice

```tsx
// A handle (in render):
const session = useIterateSession();          // suspends once, on first connect
const itx = useItx();                         // the ambient project; useItx("slug") for another

// A finite read that suspends into the router's page-level <Suspense>:
const files = useItxQuery({
  key: ["repo-files", projectId, repoPath],   // include the project id — keys are caller-owned
  query: (itx) => itx.repos.get(repoPath).listFiles(),
});

// Server-pushed live state — snapshot then diffs, never suspends:
const streams = useLiveState((itx) => itx.liveState, (s) => s.streamsIndex);

// A mutation is just a capability call (no hook):
const send = useMutation({ mutationFn: (t: string) => itx.chat.sendMessage(t) });

// The one mount, under an `ssr: false` route:
<ProjectScope slug={project.slug}><Outlet /></ProjectScope>
```

## What got deleted

- the socket `Map`, `socketKey`, `connectionKey`, `ItxAddress`, per-context keying
- `evictItxSocket` / `evictItxSocketIfCurrent`, the young-socket guard, `reconnectAllItx`
- **`<ItxProvider>`** — nothing needs a provider above the hooks; its one real job (socket pre-warm) folded into `<ProjectScope>`
- **the 12 per-page `<ItxBoundary>` wrappers** — TanStack Router already wraps every route match in `<Suspense>` (the router's `defaultPendingComponent`), so they were pure redundancy
- the `RpcStub<Session & Project>` handle intersection and the public-boundary casts (types are now honest; one documented internal cast remains at the identity boundary, bridging capnweb's `Stubify` nesting)
- the mirror's dedicated socket
- the snapshot `status`/`error` fields (nothing read them)
- *(round 2)* the `liveSession` global (the snapshot is authoritative), the `Generation.id/session/disposed` fields (object identity IS the CAS), the unused `itx:` opt threaded through three hooks (zero callers), the redundant subscription dial-retry timer (a failed dial has already published a paced successor generation — that dep is the retry), the duplicated liveness constants, and the duplicated TanStack retry lambdas

Net across the PR: **+2,271 / −1,328** over 51 files (docs are ~40% of the additions — the frontend guide, the stream-collapse design, the roadmap). `itx-react.tsx` itself went 1092 → 1185: the socket-Map apparatus left, and the generation/verifier machinery, the awaited-identity dial, the stable first-connect promise, the session-query hook, and a full model header moved in — every line of that growth is the transport model made explicit rather than implicit.

## Reviews

- **Pre-implementation plan review** — codex (gpt-5.6-sol, max reasoning) for React 19 / react-doctor / TanStack Start idiom. Verdict GO-WITH-CHANGES; corrections folded in: immutable snapshot, session-owned (not `useMemo`) project-stub cache, generation-guarded reconnect, typed transport-only query retry, project-namespaced query keys, unconditional `<ProjectScope>` read. (One suggestion — confirm-auth-before-publish — was deliberately **not** adopted: it adds a terminal/retryable branch in the hot path for a case `_app` middleware already gates; rationale in `DECISIONS.md` D24.)
- **Thermo-nuclear code-quality review** (cursor-team-kit) — by me and by codex (gpt-5.6-sol, max). Codex returned a BLOCK with 8 findings; **all fixed**: unhandled probe rejection (verifier now returns an explicit `"alive" | "strike"` verdict), per-generation `verifying` latch, publish-CAS on a late `open`, deferred session disposal (retire the prior session only when the successor publishes), narrowed `isItxTransportError`, imperative narrowing routed through `connectItx`, the REPL split into session/project branches, and an e2e assertion fix. Re-review clean.
- **Follow-up model review** — a second round (subagents + codex, gpt-5.6-sol max) on *larger* refactors toward the single programming model. The no-brainers are done here; the rest is written up in the roadmap doc below.
- **Fable gap-review of the fixes themselves** — briefed on everything the prior rounds found and asked what the fixes *broke*. It independently re-derived the terminal-auth dial storm (fixed by then — parking, the same latch direction it recommended) and found one blocker nobody else had: **`useLiveState`'s node identity missed the ambient slug**. TanStack Router doesn't remount route components on param-only navigation, so `/projects/a/repos → /projects/b/repos` re-pointed the subscription without resetting the store — project A's live state rendered under project B until B's first push (B's secrets page flashing A's secret names). The node key is now the *effective* slug (`opts.slug ?? ambient`), with a regression test driving the real navigation shape. It also had the dial-timeout fix pinned by a new hung-handshake test, de-vacuous'd the late-open CAS test (it now asserts a superseded corpse never even authenticates), and got the barrier's two accepted transient edges documented in place. Its full clean-verification list (dial state machine settle-once, firstConnect can't wedge, signal threading, retry classification uniformity) matched the implementation.
- **Second thermo round** (codex gpt-5.6-sol max + an independent second-model pass, each blind to the other) on the finished surface. Seven real bugs found, **all fixed** with regression tests: the session-identity fork + hung-authenticate wedge (above); first-load suspense reaching the error boundary on a closed-before-open dial (React `use()` thenable replay); subscription handles never `Symbol.dispose`d — an import-table leak on the now-tab-long socket (also fixed in `useLiveState`'s wrapper and the ⌘K one-shot read); permanent subscribe failures retrying forever every 10s (now transport-only retry); stale async subscribe runs flipping state after supersession (one cancellation signal threads through); a missing third capnweb transport signature (`WebSocket connection failed.`) in `isItxTransportError`; and `useLiveState` briefly showing the previous node's value when re-pointed via `opts.slug` (node-identity barrier). Plus the cuts listed above. Both rounds independently re-endorsed the two-read-hook split, the combined `ProjectScope`, and the (renamed, narrowed) private `useReconnectableItxEffect` seam.

## Trade-offs

- **Shared fault domain**: one socket, so any reconnect affects the whole tab — but reconnect is now invisible (stale-while-revalidate + no connection-hook re-suspend), which is exactly what makes the single socket safe and retires the reason the mirror was isolated.
- **A stray imperative action fired during the sub-second reconnect gap** rides the dead stub and rejects (reads await the fresh connection, so they're unaffected). Idempotency-keyed appends can retry.

## Docs

- **New: [`docs/frontend-development.md`](https://github.com/iterate/iterate/blob/three-websockets/docs/frontend-development.md)** (linked from the root README) — the apps/os frontend programming model end to end: the stack, the four nouns, one-socket/invisible-reconnect, the LiveView rules of thumb, per-hook examples, and the full 15-export table.
- **[`apps/os/docs/itx-frontend-model-roadmap.md`](https://github.com/iterate/iterate/blob/three-websockets/apps/os/docs/itx-frontend-model-roadmap.md)** — where this is going: capability-aligned live projections (so integrations/scheduler stop hand-invalidating), a non-React transport core, retiring the browser stream mirror. Also records *considered-and-kept-as-is* decisions (e.g. `useReconnectableItxEffect` stays a private single-responsibility seam).
- **NEW: [`apps/os/docs/stream-mirror-collapse.md`](https://github.com/iterate/iterate/blob/three-websockets/apps/os/docs/stream-mirror-collapse.md)** — the reviewed design for the one place the model doesn't hold yet: the stream feed's browser mirror (a second consumption model, ~4,850 LOC of OPFS SQLite + leader election + checkpoints). Collapse plan: a server-owned bounded feed live view (`itx.streams.get(path).liveState`) + cursor-paged history over the **already-existing** `getEvents` pager, phased A–D, with the honest losses spelled out. The frontend guide now documents this seam explicitly ("The one exception (for now): the stream feed") so the uniformity isn't oversold.
- `DECISIONS.md` **D24** (supersedes the socket-Map-per-context decision; records the confirm-auth-before-publish reversal), the `itx-react.tsx` module header, and the itx README file table.

## Testing

- `apps/os/src/itx/itx-react.test.ts` rewritten for the one-socket model (**27 tests**): one socket per tab, slug narrowing, invisible reconnect (live session survives a transport gap + auto-redial), the generation-guarded half-open verifier, the subscription liveness watchdog — plus the review-round regressions: first-load suspense surviving a closed-before-open dial (error boundary never hit) with the hook and imperative paths sharing ONE session identity (the capnweb mock's first `authenticate` now returns a **thenable**, modeling the RpcPromise), a hung handshake closed by the dial timeout, terminal auth rejection **parking** (repeated connects + a rendered hook leave the socket count at 1; a semantic reset revives) — including the reconnect shape (a live session's redial rejecting terminally drops + disposes the dead session instead of serving zombie stubs), unsubscribe+dispose on teardown, transport-vs-permanent subscribe retry classification, a superseded mid-subscribe run disposing its late handle without flipping state, the ambient-slug node switch dropping the held live value immediately, online-while-hidden recovery, and semantic-reset backoff clearing.
- Full apps/os unit suite green (**1,688 tests**); typecheck, oxlint (`--deny-warnings`), oxfmt clean.

No migration path / back-compat by design.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **High Risk**
> Touches the core real-time transport for the whole dashboard (shared socket fault domain, reconnect/auth parking, and stream mirror half-open handling); regressions could blank pages, leak stubs, or show cross-project live state.
> 
> **Overview**
> Collapses the apps/os browser from **three per-context WebSockets** (global catalog, project page, isolated stream mirror) to **one module-global socket per tab**, with `useIterateSession` / `connectIterateSession` as the session gate and `useItx(slug)` / `connectItx(slug)` narrowing to projects. **Invisible reconnect** keeps the last session across transport gaps (stale-while-revalidate reads; only in-flight transport failures retry), while **`reconnectIterateSession()`** remains the explicit semantic reset after new claims.
> 
> **Removed:** per-context socket `Map`, `connectionKey`, `ItxProvider`, per-page `ItxBoundary` wrappers, and direct socket eviction (`evictItxSocket*`). **Added/renamed API surface:** `useIterateSessionQuery` for non-suspending shell reads (sidebar, ⌘K); stream mirror now calls **`reportTransportSuspicion`** instead of evicting a dedicated lane. **`projects.get`** accepts URL **slug or `prj_` id** (e2e updated); shell components migrate from hand-rolled `useQuery` / `connectItxBrowser` to the new hooks and `{ slug }` on `useLiveState` outside `<ProjectScope>`.
> 
> **Docs:** root README link to `docs/frontend-development.md`, **D24** in `DECISIONS.md`, plus roadmap and stream-mirror-collapse design. **`itx-react.test.ts`** is largely rewritten (~27 tests) for generations, verifier, parking on terminal auth, and subscription/live-state edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9b2c8d5d9773f21347f4815b4737d0ee840a4398. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +1,095 -1,032 | +499 -468 🟩🟩🟩🟥🟥 |
| UI components | +98 -109 | +45 -55 🟥⬜⬜⬜⬜ |
| Tests | +606 -185 | +486 -147 🟩🟩🟥⬜⬜ |
| Docs | +727 -24 | +618 -24 🟩🟩🟩⬜⬜ |
| Generated | +22 -6 | +4 -4 🟩⬜⬜⬜⬜ |
| **Total** | **+2,548 -1,356** | **+1,652 -698** 🟩🟩🟩🟩🟥 |

<sub>Lines counts every changed line; Significant ignores blank lines and JS comments. Between 8ceb5a0 and 9b2c8d5, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `9b2c8d5` | [https://auth.iterate-preview-5.com](https://auth.iterate-preview-5.com) | 822.3 KiB | 20.4s | 11.9s |  | 2.3s | [Workflow run](https://github.com/iterate/iterate/actions/runs/108056146098425) | 2026-07-17T06:07:58.235Z | Preview app released. |
| Dummy Petshop | released | `9b2c8d5` | [https://dummy-petshop.iterate-preview-5.com](https://dummy-petshop.iterate-preview-5.com) | 201.7 KiB | 8.8s | 8.3s |  | 654ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/108056146098425) | 2026-07-17T06:07:56.586Z | Preview app released. |
| OS | released | `9b2c8d5` | [https://os.iterate-preview-5.com](https://os.iterate-preview-5.com) | 4.35 MiB (+0.5 KiB vs main) | 104.9s | 196.0s |  | 109.0s | [Workflow run](https://github.com/iterate/iterate/actions/runs/108056146098425) | 2026-07-17T06:07:54.863Z | Preview app released. |
| Semaphore | released | `9b2c8d5` | [https://semaphore.iterate-preview-5.com](https://semaphore.iterate-preview-5.com) | 440.7 KiB | 14.0s | 21.2s |  | 552ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/108056146098425) | 2026-07-17T06:06:06.414Z | Preview app released. |
| Streams Example App | released | `9b2c8d5` | [https://streams.iterate-preview-5.com](https://streams.iterate-preview-5.com) | 1.44 MiB | 20.7s | 58.4s | 1 retried: large streams stay virtualized and can scroll from tail to earliest rows (playwright x1) | 518ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/108056146098425) | 2026-07-17T06:07:56.447Z | Preview app released. |

</details>

### CODE REVIEW

File: apps/os/src/routes/admin.tsx [hunk 1]
```diff
@@ -115,7 +115,7 @@ function AdminGate() {
     return () => {
       cancelled = true;
     };
-  }, [itx]);
+  }, [session]);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/routes/admin.tsx:125]: Fixed in a0fa06e94: a transport-shaped probe rejection now keeps status `checking` (the reconnect re-probes on the fresh session via the `[session]` dep); only a genuine authority rejection locks. Uses the exported `isItxTransportError`.

File: apps/os/src/itx/itx-react.tsx [hunk 2]
```diff
@@ -93,514 +96,486 @@ import type { LiveStateRpc } from "../domains/streams/rpc-types.ts";
...
-              resolve();
-            },
-            (error: unknown) => {
-              clearTimeout(timer);
-              reject(error instanceof Error ? error : new Error(String(error)));
-            },
-          );
-        });
-      try {
-        // Two-strike, like every other liveness lane: one slow answer is a
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx:394]: Fixed in a0fa06e94: `verifying` is now a per-`Generation` latch, so a probe stuck on a retired generation can no longer block verifying its successor.

File: apps/os/src/itx/itx-react.tsx [hunk 3]
```diff
@@ -93,514 +96,489 @@ import type { LiveStateRpc } from "../domains/streams/rpc-types.ts";
...
+      Promise.race([
+        generation.ping!().then(() => true),
+        new Promise<boolean>((resolve) =>
+          setTimeout(() => resolve(false), LIVENESS_PROBE_TIMEOUT_MS),
+        ),
+      ]);
+    // Two-strike: one slow answer is a busy server, not a dead socket.
+    let alive = await probeOnce();
+    if (!alive && current === generation) alive = await probeOnce();
+    if (!alive) reconnectIfCurrent(generation);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx]: Fixed in a0fa06e94: `verifyTransport` now returns an explicit `alive|strike` verdict — a rejected probe is caught (no unhandled rejection). A transport-close rejection is a strike; an application/auth rejection counts as alive (the socket answered). Two strikes → `reconnectIfCurrent`.

File: apps/os/src/routes/_app/itx-repl.tsx [hunk 4]
```diff
@@ -85,7 +85,13 @@ function ItxReplConnected({
...
   scope?: Record<string, unknown>;
 }) {
-  const itx = useItx({ projectId: poolContext });
+  // One socket either way: the session (global REPL) or a project itx narrowed
+  // from it (`session.projects.get`, poolContext = a project id/slug). The pool
+  // owns the connection; the REPL never disposes this handle.
+  const session = useSession();
+  const itx = (poolContext === undefined
+    ? session
+    : session.projects.get(poolContext)) as unknown as ItxReactHandle;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/routes/_app/itx-repl.tsx]: Fixed in a0fa06e94: the REPL is split into session vs project branches — the project branch uses the cached `useItx(poolContext)` instead of a per-render `session.projects.get()`, typed as an honest `RpcStub<Session> | RpcStub<Project>` union (no double-cast).

**Comment by jonastemplestein** [comment]
## CI status

All checks I control are green: **Lint & Typecheck ✅, Test ✅, publish ✅, autofix ✅, LOC ✅, Continuous Releases ✅**, and all Cursor Bugbot threads resolved.

The one red check — **Cloudflare Previews / deploy + e2e** — is a **Cloudflare platform outage, not this change**. Every preview deploy is currently failing at the `wrangler deploy` step provisioning the AI Search namespace:

```
A request to the Cloudflare API (/accounts/…/ai-search/namespaces/os-preview-5) failed.
Internal Error [code: 7001]
```

This is not specific to this PR or slot: PR #2033 on **preview slot 7** fails identically on `os-preview-7`. This branch touches only the browser itx connection layer and the server-side `projects.get` slug resolution — nothing near ai-search, wrangler config, or the deploy scripts. It will go green on a re-run once the Cloudflare AI Search API recovers (the deploy fails before the itx e2e suite — including the updated `projects.get(slug)` assertions — even runs).

**Comment by jonastemplestein** [comment]
✅ **CI is fully green now** — the Cloudflare AI Search API recovered and the preview deploy + e2e passed (the itx e2e suite, including the new `projects.get(slug)` == `projects.get(id)` assertions, ran against the live preview). Every check is green.

File: apps/os/src/itx/itx-react.tsx [hunk 5]
```diff
@@ -93,514 +96,525 @@ import type { LiveStateRpc } from "../domains/streams/rpc-types.ts";
...
+      }, LIVENESS_PROBE_INTERVAL_MS);
+      // Retire the PREVIOUS published session — exactly once, now that its
+      // successor is live. It was kept alive through the reconnect gap so
+      // useSession()/useItx() never handed out a disposed stub; dispose its
+      // project-stub cache + the stub itself only here.
+      const priorSession = liveSession;
+      liveSession = root;
+      setSnapshot({ generation: id, session: root, connecting: promise });
+      resolve(root);
+      if (priorSession !== undefined) disposeSession(priorSession);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx]: Not a real defect. `disposeSession(priorSession)` only runs in the `open` handler when a successor publishes, and a successor only dials after `current` was cleared — which every path (the `close` handler, `reconnectIfCurrent`, `reconnectIterateSession`) does **after** closing the prior socket. So the prior transport is *always already closed* when disposal runs: its subscriptions are already dead (capnweb rejects every call on transport close) and the `useItxEffect` unsubscribe cleanups are catch-wrapped for exactly that. Disposal here releases already-dead local refs, never a live subscription. Strengthened the comment to state that invariant in b787c1238.

File: apps/os/src/itx/itx-react.tsx [hunk 6]
```diff
@@ -93,514 +96,525 @@ import type { LiveStateRpc } from "../domains/streams/rpc-types.ts";
...
-  for (const entry of [...socketEntries.values()]) {
-    evictItxSocket(entry.address);
+export function reconnectItx(): void {
+  const generation = current;
+  if (generation !== undefined) {
+    current = undefined; // FIRST: the close retireGeneration triggers must not auto-redial
+    retireGeneration(generation);
   }
+  dial();
+}
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx:584]: Fixed in b787c1238: `reconnectIterateSession` now resets `consecutiveDialFailures = 0` before dialing, so a deliberate new-claims reset dials immediately instead of inheriting backoff from earlier closed-before-open failures. Added a regression test — after two failures engage pacing, a semantic reset dials immediately (socket instances 2 → 3, no timer advance).

File: apps/os/src/itx/itx-react.tsx [hunk 7]
```diff
@@ -80,527 +83,572 @@ import {
...
+function SessionPrewarm() {
+  useSession();
   return null;
 }
 
 /**
- * Sets the default itx address for a subtree (and pre-warms its socket). It hands
- * down an ADDRESS, not a handle — and pre-opens the connection here so children's
- * `useItx()` resolve synchronously to the same socket. "Almost an optimization":
- * `useItx()` works without a provider (falls back to global).
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx:468]: Fixed in b787c1238, in **both** the baseline verifier and the subscription liveness watchdog: `visibilitychange` still only probes on *becoming* visible, but the `online` handler now probes unconditionally — so a backgrounded tab recovers a half-open socket the moment connectivity returns, not on next focus. Added a regression test: an `online` event while hidden re-checks a subscription (whereas a `visibilitychange` while hidden does not).

File: apps/os/src/itx/itx-react.tsx [hunk 8]
```diff
@@ -1,788 +1,892 @@
...
+          // imperative awaiters AND the suspended tree — retrying would loop.
+          // Reject BEFORE closing: the close handler's generic dial-close
+          // rejection must never mask the real error.
+          const terminal = error instanceof Error ? error : new Error(String(error));
+          reject(terminal);
+          firstConnect?.reject(terminal);
+          firstConnect = undefined;
+          current = undefined;
+          retireGeneration(generation);
+          return;
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx:414]: Real and fixed in 44e3c0176 — the exact storm you traced: the terminal path cleared `current` without recording failure, `currentSnapshot()` dials whenever the slot is empty, and the failure's own `onConnectionError` setState re-renders — each failure scheduled another (unpaced) dial. Now a terminal auth rejection **parks**: the failed generation keeps owning the slot (`Generation.failed`), so renders can never trigger a dial; its close handler is suppressed from redialing; imperative connects keep returning the same terminal failure; and revival is explicit (`reconnectIterateSession()` or a page load). The regression test drives the storm vector end to end — repeated `connectIterateSession()` calls plus a rendered `useIterateSession()` leave the socket count at 1, the error boundary shows the terminal error, and a semantic reset revives to a working session.

File: apps/os/src/itx/itx-react.tsx [hunk 9]
```diff
@@ -1,788 +1,910 @@
...
+          // generation keeps owning the slot (see Generation.failed) so a
+          // render can't storm fresh dials; publish the parked snapshot so
+          // effects observe one final generation and settle in "error".
+          const terminal = error instanceof Error ? error : new Error(String(error));
+          generation.failed = true;
+          reject(terminal);
+          firstConnect?.reject(terminal);
+          firstConnect = undefined;
+          setSnapshot({ generation: id, session: snapshot?.session, connecting: promise });
+          retireGeneration(generation);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/itx/itx-react.tsx:413]: Real — fixed in 50428e912. Keeping the last session across a gap is the invisible-reconnect move for **transport** gaps; a terminal `authenticate` rejection on the redial is an **authority** loss, and the prior session (socket already closed) can never revive — so serving it was a zombie: every call failed with transport-shaped errors while the real auth error never reached the hook path. The parked snapshot now carries `session: undefined` — `useIterateSession()` falls through to the rejected connecting promise and surfaces the terminal error at the boundary — and the dead session's stub caches are disposed on the spot. Regression test drives the full shape: live session → socket death → auto-redial → terminal rejection → prior session disposed (not handed out), error surfaces, slot stays parked, and an explicit `reconnectIterateSession()` with restored claims recovers.
